src/wayland/idle_notify/mod.rs

```diff
@@ -0,0 +1,345 @@
+type NotificationObjectId = ObjectId;
+
+type SeatMap<T> = HashMap<SeatObjectId, T>;
+type NotificationMap<T> = HashMap<NotificationObjectId, T>;
+
+/// State of ext-idle-notify module
+#[derive(Debug)]
+pub struct IdleNotifierState<D> {
+    global: GlobalId,
+    notifications: SeatMap<NotificationMap<ExtIdleNotificationV1>>,
+    loop_handle: LoopHandle<'static, D>,
+    is_inhibited: bool,
+}
+
+impl<D: IdleNotifierHandler> IdleNotifierState<D> {
+    /// Create new [`ExtIdleNotifierV1`] global.
+    pub fn new(display: &DisplayHandle, loop_handle: LoopHandle<'static, D>) -> Self
+    where
+        D: GlobalDispatch<ExtIdleNotifierV1, ()>,
+        D: Dispatch<ExtIdleNotifierV1, ()>,
+        D: Dispatch<ExtIdleNotificationV1, IdleNotificationUserData>,
+        D: IdleNotifierHandler,
+        D: 'static,
+    {
+        let global = display.create_global::<D, ExtIdleNotifierV1, _>(1, ());
+        Self {
+            global,
+            notifications: HashMap::new(),
+            loop_handle,
+            is_inhibited: false,
+        }
+    }
+
+    /// Inhibit entering idle state, eg. by wp idle inhibit protocol
+    pub fn set_is_inhibited(&mut self, is_inhibited: bool) {
+        if self.is_inhibited == is_inhibited {
+            return;
+        }
+
+        self.is_inhibited = is_inhibited;
+
+        for notification in self.notifications() {
+            if is_inhibited {
+                let data = notification.data::<IdleNotificationUserData>().unwrap();
+
+                if data.is_idle() {
+                    notification.resumed();
+                }
+
+                if let Some(token) = data.take_timer_token() {
+                    self.loop_handle.remove(token);
+                }
+            } else {
+                self.reinsert_timer(notification);
+            }
+        }
+    }
+
+    /// Is idle state inhibited, eg. by wp idle inhibit protocol
+    pub fn is_inhibited(&mut self) -> bool {
+        self.is_inhibited
+    }
+
+    /// Should be called whenever activity occurs on a seat, eg. mouse/keyboard input
+    pub fn notify_activity(&mut self, seat: &WlSeat) {
+        let Some(notifications) = self.notifications.get(&seat.id()) else { return; };
+
+        for notification in notifications.values() {
+            let data = notification.data::<IdleNotificationUserData>().unwrap();
+
+            if data.is_idle() {
+                notification.resumed();
+                data.set_idle(false);
+            }
+
+            self.reinsert_timer(notification);
+        }
+    }
+
+    /// Returns the [`ExtIdleNotifierV1`] global.
+    pub fn global(&self) -> GlobalId {
+        self.global.clone()
+    }
+
+    fn notifications(&self) -> impl Iterator<Item = &ExtIdleNotificationV1> {
+        self.notifications
+            .values()
+            .into_iter()
+            .flat_map(|seat| seat.values())
+    }
+
+    fn reinsert_timer(&self, notification: &ExtIdleNotificationV1) {
```

Comment:
I'm thinking that maybe this logic with idle timers and inhibiting could be extracted into utils? Because if the compositor wants to do something upon idle too, subject to inhibiting, it'll have to duplicate the exact same logic essentially.